lib/block_view_message.go

```diff
@@ -79,86 +178,172 @@ func (bav *UtxoView) deleteMessageMappings(message *PGMessage) {
+	for _, messagingKeyEntry := range dbMessagingKeys {
+		key := *messagingKeyEntry.MessagingPublicKey
+		// Check if we have seen the messaging key before.
+		if _, exists := messagingKeysMap[key]; !exists {
+			messagingKeysMap[key] = messagingKeyEntry
+		} else {
+			// If the messaging key was deleted in the UtxoView, we can now delete it from our map.
+			if messagingKeysMap[key].isDeleted {
+				delete(messagingKeysMap, key)
+			}
 		}
+	}
 
-		// At this point we are confident the map key is equal to the message
-		// key containing the passed-in public key so add it to the mapping.
-		messageEntriesToReturn = append(messageEntriesToReturn, viewMessageEntry)
+	for _, messagingKeyEntry := range messagingKeysMap {
+		_messagingKeyEntries = append(_messagingKeyEntries, messagingKeyEntry)
 	}
+	return _messagingKeyEntries, nil
+}
+
+// TODO: Update for Postgres
+func (bav *UtxoView) GetMessagesForUser(publicKey []byte) (
+	_messageEntries []*MessageEntry, _messagingKeyEntries []*MessagingKeyEntry, _err error) {
 
-	return messageEntriesToReturn, nil
+	return bav.GetLimitedMessagesForUser(publicKey, math.MaxUint64)
 }
 
 // TODO: Update for Postgres
-func (bav *UtxoView) GetLimitedMessagesForUser(publicKey []byte) (
-	_messageEntries []*MessageEntry, _err error) {
+func (bav *UtxoView) GetLimitedMessagesForUser(publicKey []byte, limit uint64) (
+	_messageEntries []*MessageEntry, _messagingKeyEntries []*MessagingKeyEntry, _err error) {
 
-	// Start by fetching all the messages we have in the db.
-	dbMessageEntries, err := DbGetLimitedMessageEntriesForPublicKey(bav.Handle, publicKey)
+	// First get all messaging keys for a user.
+	var err error
+	_messagingKeyEntries, err = bav.GetUserMessagingKeys(publicKey)
 	if err != nil {
-		return nil, errors.Wrapf(err, "GetMessagesForUser: Problem fetching MessageEntrys from db: ")
+		return nil, nil, errors.Wrapf(err, "GetLimitedMessagesForUser: " +
+			"problem getting user messaging keys")
 	}
 
-	// Iterate through the entries found in the db and force the view to load them.
-	// This fills in any gaps in the view so that, after this, the view should contain
-	// the union of what it had before plus what was in the db.
-	for _, dbMessageEntry := range dbMessageEntries {
-		messageKey := MakeMessageKey(publicKey, dbMessageEntry.TstampNanos)
-		bav._getMessageEntryForMessageKey(&messageKey)
+	// We define an auxiliary map to keep track of messages in UtxoView and DB.
+	messagesMap := make(map[MessageKey]*MessageEntry)
+
+	// First look for messages in the UtxoView. We don't skip deleted entries for now as we will do it later.
+	for messageKey, messageEntry := range bav.MessageKeyToMessageEntry {
+		for _, messagingKeyEntry := range _messagingKeyEntries {
+			if reflect.DeepEqual(messageKey.PublicKey[:], messagingKeyEntry.MessagingPublicKey[:]) {
+				// We will add the messages by sender so that we have no overlaps in the DB in some weird edge cases.
```

Comment:
What are these edge cases referenced here?

Reply:
This is check is for fetching messages from UtxoView and DB and it's potentially possible that the same message is present in both places. To make sure we don't fetch the same message twice, we will index messages in the local `messagesMap` by sender public key (we could alternatively index it by recipient public key, so we just choose sender).